// realm/binding/dist/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Status = exports.Float = exports.ProxyType = exports.WatchStreamState = exports.SyncSubscriptionSetState = exports.SyncUserState = exports.ProgressDirection = exports.SyncErrorAction = exports.SyncSessionConnectionState = exports.SyncSessionState = exports.SyncSessionStopPolicy = exports.HttpMethod = exports.LoggerLevel = exports.MetadataMode = exports.AuthProvider = exports.AppCacheMode = exports.DataType = exports.ClientResetMode = exports.TableType = exports.PropertyType = exports.SchemaMode = void 0;
////////////////////////////////////////////////////////////////////////////
//
// Copyright 2024 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////
var SchemaMode;
(function (SchemaMode) {
    SchemaMode[SchemaMode["Automatic"] = 0] = "Automatic";
    SchemaMode[SchemaMode["Immutable"] = 1] = "Immutable";
    SchemaMode[SchemaMode["ReadOnly"] = 2] = "ReadOnly";
    SchemaMode[SchemaMode["SoftResetFile"] = 3] = "SoftResetFile";
    SchemaMode[SchemaMode["HardResetFile"] = 4] = "HardResetFile";
    SchemaMode[SchemaMode["AdditiveDiscovered"] = 5] = "AdditiveDiscovered";
    SchemaMode[SchemaMode["AdditiveExplicit"] = 6] = "AdditiveExplicit";
    SchemaMode[SchemaMode["Manual"] = 7] = "Manual";
})(SchemaMode = exports.SchemaMode || (exports.SchemaMode = {}));
var PropertyType;
(function (PropertyType) {
    PropertyType[PropertyType["Int"] = 0] = "Int";
    PropertyType[PropertyType["Bool"] = 1] = "Bool";
    PropertyType[PropertyType["String"] = 2] = "String";
    PropertyType[PropertyType["Data"] = 3] = "Data";
    PropertyType[PropertyType["Date"] = 4] = "Date";
    PropertyType[PropertyType["Float"] = 5] = "Float";
    PropertyType[PropertyType["Double"] = 6] = "Double";
    PropertyType[PropertyType["Object"] = 7] = "Object";
    PropertyType[PropertyType["LinkingObjects"] = 8] = "LinkingObjects";
    PropertyType[PropertyType["Mixed"] = 9] = "Mixed";
    PropertyType[PropertyType["ObjectId"] = 10] = "ObjectId";
    PropertyType[PropertyType["Decimal"] = 11] = "Decimal";
    PropertyType[PropertyType["Uuid"] = 12] = "Uuid";
    PropertyType[PropertyType["Required"] = 0] = "Required";
    PropertyType[PropertyType["Nullable"] = 64] = "Nullable";
    PropertyType[PropertyType["Array"] = 128] = "Array";
    PropertyType[PropertyType["Set"] = 256] = "Set";
    PropertyType[PropertyType["Dictionary"] = 512] = "Dictionary";
    PropertyType[PropertyType["Collection"] = 896] = "Collection";
    PropertyType[PropertyType["Flags"] = 960] = "Flags";
})(PropertyType = exports.PropertyType || (exports.PropertyType = {}));
var TableType;
(function (TableType) {
    TableType[TableType["TopLevel"] = 0] = "TopLevel";
    TableType[TableType["Embedded"] = 1] = "Embedded";
    TableType[TableType["TopLevelAsymmetric"] = 2] = "TopLevelAsymmetric";
})(TableType = exports.TableType || (exports.TableType = {}));
var ClientResetMode;
(function (ClientResetMode) {
    ClientResetMode[ClientResetMode["Manual"] = 0] = "Manual";
    ClientResetMode[ClientResetMode["DiscardLocal"] = 1] = "DiscardLocal";
    ClientResetMode[ClientResetMode["Recover"] = 2] = "Recover";
    ClientResetMode[ClientResetMode["RecoverOrDiscard"] = 3] = "RecoverOrDiscard";
})(ClientResetMode = exports.ClientResetMode || (exports.ClientResetMode = {}));
var DataType;
(function (DataType) {
    DataType[DataType["Int"] = 0] = "Int";
    DataType[DataType["Bool"] = 1] = "Bool";
    DataType[DataType["String"] = 2] = "String";
    DataType[DataType["Binary"] = 4] = "Binary";
    DataType[DataType["Mixed"] = 6] = "Mixed";
    DataType[DataType["Timestamp"] = 8] = "Timestamp";
    DataType[DataType["Float"] = 9] = "Float";
    DataType[DataType["Double"] = 10] = "Double";
    DataType[DataType["Decimal"] = 11] = "Decimal";
    DataType[DataType["Link"] = 12] = "Link";
    DataType[DataType["LinkList"] = 13] = "LinkList";
    DataType[DataType["ObjectId"] = 15] = "ObjectId";
    DataType[DataType["TypedLink"] = 16] = "TypedLink";
    DataType[DataType["Uuid"] = 17] = "Uuid";
})(DataType = exports.DataType || (exports.DataType = {}));
var AppCacheMode;
(function (AppCacheMode) {
    AppCacheMode[AppCacheMode["Enabled"] = 0] = "Enabled";
    AppCacheMode[AppCacheMode["Disabled"] = 1] = "Disabled";
})(AppCacheMode = exports.AppCacheMode || (exports.AppCacheMode = {}));
var AuthProvider;
(function (AuthProvider) {
    AuthProvider[AuthProvider["Anonymous"] = 0] = "Anonymous";
    AuthProvider[AuthProvider["AnonymousNoReuse"] = 1] = "AnonymousNoReuse";
    AuthProvider[AuthProvider["Facebook"] = 2] = "Facebook";
    AuthProvider[AuthProvider["Google"] = 3] = "Google";
    AuthProvider[AuthProvider["Apple"] = 4] = "Apple";
    AuthProvider[AuthProvider["Custom"] = 5] = "Custom";
    AuthProvider[AuthProvider["UsernamePassword"] = 6] = "UsernamePassword";
    AuthProvider[AuthProvider["Function"] = 7] = "Function";
    AuthProvider[AuthProvider["ApiKey"] = 8] = "ApiKey";
})(AuthProvider = exports.AuthProvider || (exports.AuthProvider = {}));
var MetadataMode;
(function (MetadataMode) {
    MetadataMode[MetadataMode["NoEncryption"] = 0] = "NoEncryption";
    MetadataMode[MetadataMode["Encryption"] = 1] = "Encryption";
    MetadataMode[MetadataMode["NoMetadata"] = 2] = "NoMetadata";
})(MetadataMode = exports.MetadataMode || (exports.MetadataMode = {}));
var LoggerLevel;
(function (LoggerLevel) {
    LoggerLevel[LoggerLevel["All"] = 0] = "All";
    LoggerLevel[LoggerLevel["Trace"] = 1] = "Trace";
    LoggerLevel[LoggerLevel["Debug"] = 2] = "Debug";
    LoggerLevel[LoggerLevel["Detail"] = 3] = "Detail";
    LoggerLevel[LoggerLevel["Info"] = 4] = "Info";
    LoggerLevel[LoggerLevel["Warn"] = 5] = "Warn";
    LoggerLevel[LoggerLevel["Error"] = 6] = "Error";
    LoggerLevel[LoggerLevel["Fatal"] = 7] = "Fatal";
    LoggerLevel[LoggerLevel["Off"] = 8] = "Off";
})(LoggerLevel = exports.LoggerLevel || (exports.LoggerLevel = {}));
var HttpMethod;
(function (HttpMethod) {
    HttpMethod[HttpMethod["Get"] = 0] = "Get";
    HttpMethod[HttpMethod["Post"] = 1] = "Post";
    HttpMethod[HttpMethod["Patch"] = 2] = "Patch";
    HttpMethod[HttpMethod["Put"] = 3] = "Put";
    HttpMethod[HttpMethod["Del"] = 4] = "Del";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));
var SyncSessionStopPolicy;
(function (SyncSessionStopPolicy) {
    SyncSessionStopPolicy[SyncSessionStopPolicy["Immediately"] = 0] = "Immediately";
    SyncSessionStopPolicy[SyncSessionStopPolicy["LiveIndefinitely"] = 1] = "LiveIndefinitely";
    SyncSessionStopPolicy[SyncSessionStopPolicy["AfterChangesUploaded"] = 2] = "AfterChangesUploaded";
})(SyncSessionStopPolicy = exports.SyncSessionStopPolicy || (exports.SyncSessionStopPolicy = {}));
var SyncSessionState;
(function (SyncSessionState) {
    SyncSessionState[SyncSessionState["Active"] = 0] = "Active";
    SyncSessionState[SyncSessionState["Dying"] = 1] = "Dying";
    SyncSessionState[SyncSessionState["Inactive"] = 2] = "Inactive";
    SyncSessionState[SyncSessionState["WaitingForAccessToken"] = 3] = "WaitingForAccessToken";
})(SyncSessionState = exports.SyncSessionState || (exports.SyncSessionState = {}));
var SyncSessionConnectionState;
(function (SyncSessionConnectionState) {
    SyncSessionConnectionState[SyncSessionConnectionState["Disconnected"] = 0] = "Disconnected";
    SyncSessionConnectionState[SyncSessionConnectionState["Connecting"] = 1] = "Connecting";
    SyncSessionConnectionState[SyncSessionConnectionState["Connected"] = 2] = "Connected";
})(SyncSessionConnectionState = exports.SyncSessionConnectionState || (exports.SyncSessionConnectionState = {}));
var SyncErrorAction;
(function (SyncErrorAction) {
    SyncErrorAction[SyncErrorAction["NoAction"] = 0] = "NoAction";
    SyncErrorAction[SyncErrorAction["ProtocolViolation"] = 1] = "ProtocolViolation";
    SyncErrorAction[SyncErrorAction["ApplicationBug"] = 2] = "ApplicationBug";
    SyncErrorAction[SyncErrorAction["Warning"] = 3] = "Warning";
    SyncErrorAction[SyncErrorAction["Transient"] = 4] = "Transient";
    SyncErrorAction[SyncErrorAction["DeleteRealm"] = 5] = "DeleteRealm";
    SyncErrorAction[SyncErrorAction["ClientReset"] = 6] = "ClientReset";
    SyncErrorAction[SyncErrorAction["ClientResetNoRecovery"] = 7] = "ClientResetNoRecovery";
})(SyncErrorAction = exports.SyncErrorAction || (exports.SyncErrorAction = {}));
var ProgressDirection;
(function (ProgressDirection) {
    ProgressDirection[ProgressDirection["Upload"] = 0] = "Upload";
    ProgressDirection[ProgressDirection["Download"] = 1] = "Download";
})(ProgressDirection = exports.ProgressDirection || (exports.ProgressDirection = {}));
var SyncUserState;
(function (SyncUserState) {
    SyncUserState[SyncUserState["LoggedOut"] = 0] = "LoggedOut";
    SyncUserState[SyncUserState["LoggedIn"] = 1] = "LoggedIn";
    SyncUserState[SyncUserState["Removed"] = 2] = "Removed";
})(SyncUserState = exports.SyncUserState || (exports.SyncUserState = {}));
var SyncSubscriptionSetState;
(function (SyncSubscriptionSetState) {
    SyncSubscriptionSetState[SyncSubscriptionSetState["Uncommitted"] = 0] = "Uncommitted";
    SyncSubscriptionSetState[SyncSubscriptionSetState["Pending"] = 1] = "Pending";
    SyncSubscriptionSetState[SyncSubscriptionSetState["Bootstrapping"] = 2] = "Bootstrapping";
    SyncSubscriptionSetState[SyncSubscriptionSetState["Complete"] = 3] = "Complete";
    SyncSubscriptionSetState[SyncSubscriptionSetState["Error"] = 4] = "Error";
    SyncSubscriptionSetState[SyncSubscriptionSetState["Superseded"] = 5] = "Superseded";
    SyncSubscriptionSetState[SyncSubscriptionSetState["AwaitingMark"] = 6] = "AwaitingMark";
})(SyncSubscriptionSetState = exports.SyncSubscriptionSetState || (exports.SyncSubscriptionSetState = {}));
var WatchStreamState;
(function (WatchStreamState) {
    WatchStreamState[WatchStreamState["NeedData"] = 0] = "NeedData";
    WatchStreamState[WatchStreamState["HaveEvent"] = 1] = "HaveEvent";
    WatchStreamState[WatchStreamState["HaveError"] = 2] = "HaveError";
})(WatchStreamState = exports.WatchStreamState || (exports.WatchStreamState = {}));
var ProxyType;
(function (ProxyType) {
    ProxyType[ProxyType["Http"] = 0] = "Http";
    ProxyType[ProxyType["Https"] = 1] = "Https";
})(ProxyType = exports.ProxyType || (exports.ProxyType = {}));
// Wrapped types
class Float {
    value;
    constructor(value) {
        this.value = value;
    }
    valueOf() {
        return this.value;
    }
}
exports.Float = Float;
class Status {
    isOk;
    code;
    reason;
    constructor(isOk) {
        this.isOk = isOk;
    }
}
exports.Status = Status;
